### PR TITLE
[Graphbolt] Support converting dgl node classification dataset to graphbolt.

### PR DESCRIPTION
**Pull request by frozenbugs**
## Description

Tested: RedditDataset, cora, ogbn-mag, ogbn-products, ogbn-arxiv.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tests/python/pytorch/graphbolt/impl/test_legacy_dataset.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import dgl.graphbolt as gb
+import pytest
+import torch
+from dgl import AddSelfLoop
+from dgl.data import AsNodePredDataset, CoraGraphDataset
+
+
+def test_LegacyDataset_homo_node_pred():
+    cora = CoraGraphDataset(transform=AddSelfLoop())
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/impl/test_legacy_dataset.py:9]: For ogb datasets, is it possible to transform the graph when generating gb.Dataset?

**Review by Rhett-Ying — COMMENTED** [review]
I am wondering how about converting various datasets to `DGLGraph`(as an intermediate representation or bridge) first, then compose `gb.Dataset` from the `DGLGraph`?

File: python/dgl/graphbolt/impl/legacy_dataset.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            tensor = legacy[0].ndata[name]
+            if tensor.dim() == 1:
+                tensor = tensor.view(-1, 1)
+            features[("node", None, name)] = TorchBasedFeature(tensor)
+        for name in legacy[0].edata.keys():
+            tensor = legacy[0].edata[name]
+            if tensor.dim() == 1:
+                tensor = tensor.view(-1, 1)
+            features[("edge", None, name)] = TorchBasedFeature(tensor)
+        self._feature = BasicFeatureStore(features)
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply python/dgl/graphbolt/impl/legacy_dataset.py:129]: https://github.com/dmlc/dgl/blob/c213444268aa618c3bef988f74c836e10646e6e2/python/dgl/graphbolt/impl/torch_based_feature_store.py#L207

The initiation method is defined for on-disk

File: python/dgl/graphbolt/impl/legacy_dataset.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    """A Graphbolt dataset for legacy DGLDataset."""
+
+    def __init__(self, legacy: DGLDataset):
+        # Only supports single graph cases.
+        assert len(legacy) == 1
+        graph = legacy[0]
+        # Handle OGB Dataset.
+        if isinstance(graph, tuple):
+            graph, _ = graph
+        if graph.is_homogeneous:
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply python/dgl/graphbolt/impl/legacy_dataset.py:25]: Will do for `link prediction dataset` later.
I don't plan to support multiple tasks, I glanced, and did not see a lot multiple tasks dataset.

File: python/dgl/graphbolt/impl/legacy_dataset.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            for etype in graph.canonical_etypes:
+                for name in graph.edges[etype].data.keys():
+                    tensor = graph.edges[etype].data[name]
+                    if tensor.dim() == 1:
+                        tensor = tensor.view(-1, 1)
+                    gb_etype = etype_tuple_to_str(etype)
+                    features[("edge", gb_etype, name)] = TorchBasedFeature(
+                        tensor
+                    )
+            self._feature = BasicFeatureStore(features)
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply python/dgl/graphbolt/impl/legacy_dataset.py:81]: ditto.

**Comment by frozenbugs** [comment]
> I am wondering how about converting various datasets to `DGLGraph`(as an intermediate representation or bridge) first, then compose `gb.Dataset` from the `DGLGraph`?

The challenge is not on converting DGLGraph to gb graph, but create tasks, in which different dataset are very different.

**Review by Rhett-Ying — APPROVED** [review]

**Pull request merged**